File: php/signature.php
<?php /*
 * Copyright (C) 2018 All rights reserved.
 *   
 * @File signature.php
 * @Brief 
 * @Author abelzhu, abelzhu@tencent.com
 * @Version 1.0
 * @Date 2018-11-07
 *
 */
 

class SignatureGetter { 

	public function __construct($secret) 
	{ 
		$this->secret = $secret;
	}

	public function Get(string $json)
	{ 
		$jsonArray = json_decode($json, true);

		$paramArray = array();
		$this->Json2Array($jsonArray, $paramArray);

		$string = $this->ParamArray2String($paramArray);

		return $this->HmacSha256($this->secret, $string);
	}

	private function ParamArray2String(array $array)
	{
		asort($array); // sort

		//
		$str = "";
		foreach($array as $k => $v) { 
			if (0 === strpos($v, 'sig=')) { // note: do not use ==, fuck !
				continue; // escape sig self
			} 
			$str = ($str . $v . '&');
		} 
		return substr($str, 0, strlen($str)-1); // remove the last &
	}

	private function Json2Array(array $jsonArray, array &$paramArray)
	{ 
		foreach($jsonArray as $k => $v) { 
			if (is_array($v)) { 
				$this->Json2Array($v, $paramArray);
			} else { 
				array_push($paramArray, $k . '=' . $v);
			}
		}
	}

	private function HmacSha256(string $key, string $data)
	{
		return base64_encode(hash_hmac("sha256", $data, $key, $raw_output = true));
	}

	private $secret = null;
} // class

File: python/signature.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
##
 # Copyright (C) 2018 All rights reserved.
 #   
 # @File signature.py
 # @Brief 
 # @Author abelzhu, abelzhu@tencent.com
 # @Version 1.0
 # @Date 2018-11-08
 #
 #

import string
import random
import hashlib
import time
import struct
import sys    
import hmac
import base64
import json
 
class SignatureGetter(object):
    def __init__(self, secret):
        self.secret = secret

    def Get(self, jsonStr) : 
        jsonDict = json.loads(jsonStr)

        paramList = []
        self.Json2List(jsonDict, paramList)

        paramList.sort()

        return self.HmacSha256(self.List2String(paramList))

    def List2String(self, paramList):
        str = ""
        for param in paramList:
            if 0 == param.find('sig='):
                continue # egnore sig self
            str += (param + '&')
        return str.rstrip('&')


    def HmacSha256(self, data): 
        data = bytes(data).encode('utf-8')
        secret = bytes(self.secret).encode('utf-8') 
        signature = base64.b64encode(hmac.new(secret, data, digestmod=hashlib.sha256).digest())
        return signature 

    def Json2List(self, jsonDict, paramList): 
        for k, v in jsonDict.items():
            if isinstance(v, list):
                for item in v:
                    self.Json2List(item, paramList)
            else :
                paramList.append(k + '=' + str(v))
